### DISCUSSION TITLE
Slider scoring mechanics

### DISCUSSION THREAD

**Discussion by peppy** [ppy/osu#25739] [Q&A]
After somewhat heated discussion over the last few days, I'd like to create a new thread to gather any feedback on slider mechanics here.

https://twitter.com/ppy/status/1729828530651885569

https://github.com/ppy/osu/assets/191335/22ae5e7f-25ed-4bc0-b715-b5df8be0b811

From the twitter poll, the majority prefers `current`, although this comes with a disclaimer that we found out that in the video `current (miss tail)` is affected by a bug and should show the **same accuracy** as `current (miss tick)`.

From discussions in the dev discord, some users prefer `v2-1`, but I haven't seen much rational reasoning behind this preference.

To set a standard for feedback in this thread, please post a single reply with your own personal feedback (importantly: not group aggregated – have each person post for themselves).

Clearly define your feedback as based on **feel** based (how it looks and feels), **scoring** (how it affects score / acc) or other.

**Comment by DarkbitNetwork** [comment]
I prefer `v2-1` over the other implementations, but I think it'll be nicer if `v2-1`'s **visible** judgement at the slider end isn't present but mechanics are kept the same, this is because seeing two visible judgments from one slider doesn't feel right because of the same reason I have with Bubbleman (see [discord message](https://discord.com/channels/188630481301012481/188630652340404224/1183729985794625619))

Anyway, here's my reasoning for why I prefer `v2-1` over others:
1. Missing slider end drop accuracy which is why I already think `v2-1` or `v2-2` is better than `current` implementation. (See https://github.com/ppy/osu/discussions/25452) But this may be disregarded due to a bug.
2. `v2-2` punishes too much both accuracy and score-wise when compared to `v2-1`. You get punished for playing the slider wrong (getting 100 head, missing slider end, or slider tail) much harder than hitting 100 on hit circle.
This makes me feel like it'll be unnecessarily harder to get good scores on "tech maps" that heavily rely on sliders. This also makes sliders way too important compared to hit circles which I believe won't feel very good if you're coming from stable (though I only played lazer exclusively for nearly 12 months so I am not in a position to speak for players coming from stable)
2.1. Another reason is that getting punished accuracy-wise harder for aiming mistakes than rhythm mistakes simply doesn't feel right in a rhythm game. I believe it'll just lead to players prioritizing hitting slider tick and slider end over tapping slider to the beat to try to get good accuracy.
3. `Current` implementation doesn't punish missing slider tick and slider end enough compared to other implementations. Although missing slider tick already breaks the combo, I believe missing the slider end or tail should punish your accuracy (and score) more which is why I think `v2-1` and `v2-2` are better than `current`. 
3.1. If you don't punish accuracy and score for missing slider end more then this may cause "cheesing" slider end to be better than actually playing the slider properly depending on the situation. Also not punishing the slider end as much as hitting 100s on hit circle wouldn't be similar to what is currently on stable.
3.2. Missing the slider tick should also punish you as much as hitting 100s on hit circle because I don't think it makes sense for breaking combo to only reduce accuracy as much as missing a slider end on `current`. Plus, missing a slider tick also gives you 100s on stable so why not?

**Reply by felipemarins** [reply]
~Overall I agree with you about slider ends needing more weight on score. But I don't know what to feel about slider tick affecting more the accuracy as breaking combo is already a great punishment in the current score system. And I feel like accuracy is more like a metric to measure the precision you have to hit the hit circle/slider head in time, so having something like the slider tick, that only requires you to continue following the slider correctly to hit it, affecting more the accuracy seems somewhat wrong to me.~
Forget about what I said. I gave it a little more thinking and it actually makes sense as missing slider ends on `v2-1` also gives you a 100 judgment, so it would be confusing if slider ticks didn't follow the same logic.

**Reply by Zyfarok** [reply]
> I think it'll be nicer if v2-1's visible judgement at the slider end isn't present but mechanics are kept the same, this is because seeing two visible judgments from one slider doesn't feel right

It seems like the worst possible idea. If there's no judgement visible at the end, then you have no way to know whether you did hit the slider-end of or not, don't you ?
Having two judgements might need some time getting used to but it's the best as it separates the indication about the accuracy of the head from the slider ticks/ends aim requirements.

In application, most player will skin the 300 as invisible both for the head and tail, so it will only display something when you made a mistake, and while today it's almost impossible to know if you miss-acced or miss-aimed the slider-end, with that it'll be possible.

**Reply by DarkbitNetwork** [reply]
> That's the worst possible idea. If there's no judgement visible at the end, then you have no way to know whether you did hit the slider-end of or not, don't you ?

I'll say this is not the worst possible idea. This is because you can rely on other visual clues such as follow circle expanding to know when you miss a slider end like I (and Bubbleman) did, the reason why I want the judgement at slider end to not be visible is that it looks much cleaner.

Though I have never tried playing with two judgements, most people agree that having two visible judgements for one slider is confusing or unnatural and that's why I suggest having only one visible judgement

**Reply by Zyfarok** [reply]
> you can rely on other visual clues such as follow circle expanding to know when you miss a slider end

Maybe, idk. It depends on how reliable that feedback is. Because it's considered valid to leave the follow-circle up to 36ms before the slider-end, so wouldn't that just look the same as a slider-end miss even though you validated the slider ?
In any case, I agree that the feedback could be made completely separate from the presence of a judgement.

**Comment by felipemarins** [comment]
_I already gave part of my opinion on discord, but I will try to write a more complete feedback. Just a disclaimer: I'm not a developer and I'm not a scoring expert either. So this will be only based on how it **feels** and also based on some previous rhythm games I played._

In general, I feel like there should be a lot more importance in hitting the start of a "note" rather than anything else. It's how it works on most rhythm games, and some games don't even let you try to catch a long note in the middle if you didn't hit the start. I'm not saying that osu! should follow what other games do, it's just that osu! also works this way for the most part as circles are generally on the start of a "note" (may it be following a percussive instrument or something else). So letting sliders working differently should feel inconsistent, especially for new players. For a new player, it would feel strange to hit the slider head perfectly but get nearly half the accuracy just because you let go from the slider a little sooner. So in my vision, it should discard `v2-2`.

That leaves me with `current` and `v2-1`. As for `v2-1`, maybe I'm a little biased about it since I'm a lazer-only player that have been playing for almost 2 years, but it feels strange to get a judgment at the end of a slider. Maybe that's just because I got used to how sliders works in lazer, but getting two judgments on the same hit object feels unnatural to me. I also have some reservations about how it removes some weight from the slider head, as you get nearly the same accuracy if you get a 100 on slider head or if you just miss the slider tail. My suggestion on discord was to limit the maximum of the slider judgment by the judgment of the slider head while still being separated. But I guess that may just be confusing.

As for `current`, it already feels just fine to me. I liked it better when slider tails didn't give combo, but as it won't come back, I feel like there should be a better visual feedback when you miss the slider tail. Currently, I only know if I missed a slider tail if I don't hear a hitsound or on results screen if I don't have any misses on the map and I still didn't FC.

Maybe one big advantage that `v2-1` has over `current` is that it puts more weight on slider tail misses. On some other rhythm games, long notes continuously increases your score, so it would feel natural for new players to have sliders giving more score. But even if I agree that slider tails should have more weight, somehow I kind of prefer `current`. Maybe that's just because I got used to it and think that having judgments at slider ends feels so weird. I don't have much of a problem with `v2-1` itself, my only strong opinion is on `v2-2`. So if someone have strong points about preferring `v2-1` over `current`, I don't see it as much of a problem, as long as you can disable the visual judgments at the slider end and have some other visual feedback to clearly state that you missed a slider tail.

Edit: Just clarifying that I don't think slider tails are unimportant.

**Reply by Zyfarok** [reply]
> I feel like there should be a lot more importance in hitting the start of a "note" rather than anything else.

Totally agree on that point. That's the reason why I think 300 is too much for the body judgement in v2-1. However I think v2-1 with the body judgement value reduced would be the best. I suggested a value of 100 for ticks/ends and 100 for the body judgement, but 50 for ticks/ends and 150 for body might work too.

Note that in v2-1, if you don't want to see the final judgment, you can just "skin it out".

**Reply by felipemarins** [reply]
>I suggested a value of 100 for ticks/ends and 100 for the body judgement, but 50 for ticks/ends and 150 for body might work too.

Yeah, that may work out well, as the judgments in lazer are only described as "Great", "Ok" and "Meh", so you could get a "Great" for a slider and it doesn't automatically tell if it's a 300 or 150. Or you could make a different judgment just for sliders. A concern that I have is that I can't tell how much it would affect accuracy as I don't know much about how scoring works. Another concern would be about legacy skins, as some old maps have those integrated. They only have 300s, 100s and 50s, so it would be weird visually. And if you added some new judgment for sliders, those skins would have no way to infer what to do with that, unless you can fallback to the old judgments, which would still be weird. My suggestion was to cap the slider judgments by the slider head judgment while still being separated (so you get a 100 for slider judgment if you get 100 on slider head). That could be confusing but I feel like it's a little simpler to understand scoring-wise and about how it translates to legacy skins.

>Note that in v2-1, if you don't want to see the final judgment, you can just "skin it out".

If that's true, then I have no problems with `v2-1`. My only request is that there should be another way to tell visually if you miss a slider tail (e.g. setting the slider end outline to be red when you miss it, similar to when you miss the slider tick).

**Reply by Zyfarok** [reply]
Effect on accuracy is proportional to the value of the judgement. Accuracy is ratio of the sum judgement values obtained over max judgement values obtainable.

For the skin on old maps (and old skins in general), that's a good point, but I guess it could just take the 300/miss skin by default ? Or be skinned-out ? idk but it's a relatively edge-case thing anyway.

**Reply by felipemarins** [reply]
>Effect on accuracy is proportional to the value of the judgement. Accuracy is ratio of the sum judgement values obtained over max judgement values obtainable.

Thanks for the simple explanation. So I guess having a new slider judgment with different values could work well if well explained with the results screen showing the slider judgments separated from circle judgments.

>it's a relatively edge-case thing anyway.

Personally I don't care much about having legacy skins using the default lazer skin as a fallback when showing those new judgments or just using the old judgments even though being imprecise. I only mentioned it because some people might really care about it and report it as a lazer bug instead of being intentional.

**Comment by Flamiii** [comment]
<s>Personally, as someone who has played both osu!stable and osu!(lazer) for several years, I would vastly prefer `v2-2` over `current` and `v2-1` and I have a few reasons why, both feel and scoring related. Also, I'm writing all of this taking into account the `current (miss tail)` accuracy bug so keep that in mind.

Feel-wise:
1. I fundamentally dislike `v2-1` because having two separate judgments for one slider is confusing and makes no sense.
2. My biggest gripe with `current` as someone who plays on osu!(lazer) is the lack of feedback provided to me when I miss a slider tail or tick. I am someone who plays with my HUD off and there are many people who play like this. When I miss a slider tail or tick, there is no 100 or Ok at the end of the slider, so how am I supposed to know that I missed it? There is a slight red animation for missing a slider tick but it is very easy to miss, and there is no animation at all when missing a slider tail. There should be immediate feedback if I miss a slider tail or tick but `current` doesn't have that, which in my opinion, is one of the biggest issues with osu!(lazer) gameplay at the moment. 

Scoring-wise:
1. On a fundamental level, I strongly disagree with getting a 300 or a Great on a slider if you miss either the slider tail and/or tick. I don't think that this is a controversial opinion, especially among those who play osu!stable.
2. The accuracy penalty for `current` when missing a slider tail or tick might seem fine but as someone who has played osu!(lazer) prior to slider tails having bugged accuracy, I can say that the accuracy penalty is far too low for missing a slider tail or tick. When actually setting scores, slider tails and ticks count for barely anything and I feel that it is way too forgiving, especially on certain maps where your ability to follow sliders accurately is where the difficulty of the map comes from ([Example 1](https://osu.ppy.sh/beatmapsets/1137778#osu/2376713) / [Example 2](https://osu.ppy.sh/beatmapsets/2091780#osu/4384637)).

I would assume that if you were to ask higher level players who play tech and slider-heavy maps on osu!stable which version of slider scoring mechanics they prefer, the vast majority of them would probably say `v2-2`, as it is the most similar to osu!stable. I'm against `current` and `v2-1` because I feel that they potentially run the risk of invalidating leaderboards on tech and slider-heavy maps as they would make these maps much easier on osu!(lazer) than they are on osu!stable. And to be honest, I actually think osu!stable already handles slider scoring mechanics in a way that makes a lot of sense so I don't think that it's something that needs any significant changes.</s>

Edit: After seeing some other ideas being discussed here, I would advocate for Bubbleman's modified version of `v2-1` over any of the other proposals, because it addresses all of the criticisms that I initially had about `v2-1`. Bubbleman's version only has one 300/100/50/0 judgment per slider instead of two, which I find to be better than `v2-1`, and it instead has separate slider-specific judgments to keep track of things like slider ticks and slider ends. I also had an issue with `v2-1` regarding there being no feedback when missing a slider end, but he aims to have some sort of skinnable element in the game that displays when you miss a slider end, so that solves that issue as well. I can see the benefits of a system like this over `v2-2`, as it allows for all of the parts of a slider (slider head, slider ticks, and the slider tail) to be judged individually instead of them all being combined into one 300/100/50/0 judgment, and I think that is strictly better than `v2-2`, so I'm a fan of it.

![image](https://github.com/ppy/osu/assets/74088315/036f5757-6e11-4f8b-bc39-8d004fd4b16c)

**Reply by Zyfarok** [reply]
Feel-wise: I believe it should be possible to have an option to make v2-1 look like v2-2, but I'm pretty sure it's just that you're not used to v2-1 yet. I've played stable in score v2 a lot and not much lazer, but I feel like v2-1 just makes more sense because in stable v2 you don't have any good way to distinguish between a miss-acc of the head and a miss-aim of the tail.

Scoring-wise: I don't agree with your first point, it just makes sense to separate the scoring of the slider-head from the body to me, but I don't have a strong opinion and would be fine with having it be like v2-2 in that regard.

Also, don't assume what top players would answer cause from what I observed v2-1 was the dominating opinion among top players.

**Reply by Flamiii** [reply]
<s>I think my main issue with `v2-1` comes down to the fact that it's not immediately obvious which judgment(s) you would get in certain scenarios. With `v2-1` in general, does each slider give two judgments for the start and the end or just one? And if it just gives one, which one is the one that is counted? For example, if I hit a 300 on the slider head but I miss the slider tail and get a 100 at the end of the slider, what judgment do I get, the 300 or the 100? And what about the opposite scenario where I hit a 100 on the slider head and hit the rest of the slider perfectly? The fact that something as essential as how sliders are judged isn't immediately obvious is a massive issue, and I doubt it's just me. How is a new player supposed to make sense of any of this?

In terms of osu!stable ScoreV2 / `v2-2`, on slower maps, you can tell if you hit a slider head too early or late before the slider is over by looking at the UR bar, and on faster maps, you can tell through process of elimination that if you hit the body of the slider correctly but still got a 100 at the end, you probably mis-tapped the slider head. It's far more natural and intuitive than it sounds.

As for judgment counts, I think it's best for `v2-1` sliders to count for only one judgment instead of two, because otherwise judgment counts on maps won't match their object counts and it would also lead to the same map having two different judgment counts based on potentially the Classic mod or whether you are playing osu!stable or osu!(lazer), which definitely sounds like something that should be avoided if at all possible. But I also think that if `v2-1` sliders are going to show two judgments, it would be unintuitive if one of them didn't count and I would consider it to be bad design. So in my eyes, you either end up with unnecessary complexity or unintuitive behavior. And all of this is so that people can see whether or not they hit a slider head properly like half a second earlier? It feels like this change is being made to try to fix a problem that doesn't actually exist, and I don't think that this is a good direction to go in at all.

I also stand by the fact that I personally think that you shouldn't get a 300 or a perfect judgment on a slider if you don't hit all of the parts of a slider properly. I will gladly die on that hill. Also, I meant top 'tech' and slider-aim players, not just top players, as top 'tech' players are the ones most affected by a change like this ([for example, the players on this leaderboard](https://osu.ppy.sh/beatmapsets/1357624#osu/2809623)). </s>

Edit: See the edit of my initial comment above, as I don't agree with this anymore and all of these issues are addressed in Bubbleman's modified version of `v2-1`.

**Reply by Zyfarok** [reply]
> For example, if I hit a 300 on the slider head but I miss the slider tail and get a 100 at the end of the slider, what judgment do I get, the 300 or the 100?

You would get both. In the score-screen, you would get a 300 for the head, but the 300/100/50 at the end would be a separate counter I think ? At least it was discussed that it would be better to separate all "accuracy" related judgements from slider-judgements. I think it would also be better if it was named and displayed differently than the head judgements.

**Reply by Flamiii** [reply]
Do you mean something like this? So if you hit a 300 on the slider head it would count as a 300 respectively, but missing any slider ticks or ends would be under a separate category of judgments like "Slider Tick" or "Slider End"? osu!(lazer) used to do this up until a couple weeks ago and I mostly liked it. My only complaints were that slider ticks / ends didn't have enough weight accuracy-wise which can be changed pretty easily, and that there was no visible judgment for missing a slider end, but that can also be fixed by potentially having another skinnable element for it or something along those lines though I'm not entirely sure.

If this is what is being proposed, I'm totally fine with this, it just seemed very different than `v2-1`. Something like this has added benefits like being able to tell if a play has a slider break or if the player just missed a ton of slider ends because that information is on the results screen, which is something you can't do in osu!stable or in `v2-2`.

![image](https://github.com/ppy/osu/assets/74088315/3063b475-adde-4240-9be5-bc162f79e6b1)

**Comment by Bubbleman532** [comment]
The introduction of slider acc as a gameplay mechanic changes the way osu is played. Something that should have been present from the start, maps have been built around that you can tap early or late and not worry about acc on sliders, which now is changing fundamentally going forwards. Originally, since the head was just a part of the slider, the judgement coming at the end of a slider made sense, but now that the head is its own circle, scoring needs to be shaken up.

Which is what **Current** fails to do - with the head behaving as a circle, the majority of the score from a slider will happen on that hit. With what is likely **95%+ of sliders in >3 star ranked maps being just the head and a tail**, losing 1 combo and 30 score from missing the tail is a _**tiny**_ punishment compared to what I can assume is the current scorev2 method of 2-2, where missing the tail cuts your score down to a third of a well-timed hit.

However, where **2-2** fails is that it **reduces the reward for accuracy on the head** (rhythmic accuracy) **if you miss the tail** (movement accuracy). If each object with a timing accuracy is to be judged, there needs to be a concession here towards the head giving a true judgement of 300/100/50, since the tail is a fixed point in time and does not pertain to rhythmic accuracy in the same way.

This is where **2-1** has the strongest case for use, though I still believe it's fundamentally flawed since **sliders would be worth double** what they are currently in stable. High ranking players share this view. The other flaw is that sliders would **give two judgements despite being one object**, and I have a large desire to see judgement count being equal to object count - that is to say, everything you have to tap. This isn't _necessary_ but would represent rhythmic accuracy fairly on the results screen, since you could then create a static definition of full combo as "hitting every object (ticks included)" and perfect combo as "hitting _the entirety_ of every object".


Our discussions were reaching a point where Current was favoured if and only if slider ends could be worth a more arbitrary amount of points, deviating from the standard 30 for a tick, and our discussed numbers were landing anywhere from 120-200 points. We were told this was not possible, so would settle for 2-1 as a most-preferred if the slider judgement was kept separate from accuracy judgements.

As for missing slider ends, I'm now used to seeing the slider follow circle expansion animation and registering it as a missed slider end, though would still prefer some skinnable element for feedback on a hit or a miss. I don't think I'd use it personally and would skin it out, but it would greatly benefit those who need it.

**Reply by Zyfarok** [reply]
Fully agree. Only comment is that in terms of scoring value, I still believe a slider-body judgement is beneficial though, as it allows to punish the first missed tick/end more while not punishing too much if you miss the full slider, but the value of this slider-body judgement should be much less than 300. 100~150 seems to be reasonable.

**Reply by Bubbleman532** [reply]
If you miss a tick, you lose combo - that's enough of a punishment in itself.

**Reply by Zyfarok** [reply]
Yes, good point, but accuracy has (a bit) more weight than before AND if you have no combo in that section then the break has little impact. I believe it would be ok without the slider-body judgement, I'm just saying it seems a bit better with it.

**Reply by Zyfarok** [reply]
Also, this is completely subjective but to me the body judgement represents something different than the ticks/ends, which is holding the key for the full length and following the "flow" till the end. Also having a "slider-body" counter on the score screen would allow you to know how many of the sliders did you hold perfectly, regardless of the amount of ticks/ends of each.

**Reply by Bubbleman532** [reply]
This overcomplicates everything, and also ignores that on long sliders you can just release your key and cursor dance between ticks. or even between the head and slider end if there are no ticks.

**Reply by smoogipoo** [reply]
> We were told this was not possible, so would settle for 2-1 as a most-preferred if the slider judgement was kept separate from accuracy judgements.

It's not impossible.

**Reply by okolenmi** [reply]
"Hit accuracy" and "Aim accuracy" can be implemented separately, but I'm not sure peppy will accept such a big change. _It can be an extra "osu!test" gamemode like taiko/mania for people who want to test it._

**Hit accuracy** is only hit accuracy
Hit will be estimated even if cursor isn't on an object but extra hits will be **punished**.

> - No punishment for less than hitobjects_count divided by 150 for example (10 extra hits will be completely ignored for 1500 hitobjects map). Extra hits should be ignored during beatmap pauses / on spinners / on sliders partially.

**Aim accuracy** will show how far from a circle was cursor at the moment of perfect hit
For example it can be like this: inside of a circle object is 300, 1.2x diameter = 100, 1.5x diameter = 50.

**Reply by Zyfarok** [reply]
@okolenmi you totally misunderstood what was meant. Bubbleman only meant to say that since slider ticks and tails do not require the timing accuracy of a tap from the player, they should not affect the count of 300/100/50 (and by extension misses ?) on the result screen as those reflect the timing accuracy of the player on hit-circles and slider-heads.
`current` is good in that regard, but `v2-1` and `v2-2` are not, unless these 300/100/50 are counted separately.

**Comment by Zyfarok** [comment]
(Edit: I totally agree with bubbleman, I'm mostly just phrasing things differently and the only difference worth mentioning is that bubbleman seems to want to get rid of the slider-body judgement which I don't mind.)

I feel like visual feedback doesn't need to be attached to how the slider elements are valued in the scoring. They are separate issues that should be looked at separately.

### Visual feedback
In terms of visual feedback, what matters is that there's a good way to know and distinguish between if you miss-acced the head or miss-aimed ticks/ends. For miss-aimed ticks/ends however, it doesn't necessarily have to be shown through judgements.

For that purpose `v2-1` wins but `current` might be fine too:

**For the head accuracy**, stable people might want to have the option to display things as `v2-2` because it's familiar but it seems more logical to have instant feedback as in `current` or `v2-1`. This instant feedback kinda felt like it was missing in stable ScoreV2.

**And then for the ticks/ends** it depends of how easy is it to tell through the other visual clues, but `v2-1` is the safest because then people can choose through skinning whether they want the additional tail judgement. (But if `current` allows to make it clear that you missed the tail through other clues, then `current` works well too)

### Scoring value
The BaseValue of judgements ultimately impacts how much value each element has in the score, especially since the scoring is now giving more weight to accuracy than before.
On that point I have two strong intuitions that I believe most people would easily agree with:
- The total value of a slider with 0 ticks (huge majority of sliders as Bubbleman mentions) should be approximately 450~500 as it should clearly be more than the 300 of a single hit-circle but still less than two hit-circles (600) as it is strictly easier than hitting two hit-circles. In that regard, `v2-1` seems to give way too much value to the slider while `v2-2` and `current` seems a bit too low in total value.
- The slider head has no reason to be treated differently that normal hit-circles, so it makes sense to give a value of 300 on the head judgement, and about 150~200 for sum of body + tail judgements. In that regard,  `current` make more sense but the values of ticks/tails/body NEED to be adjusted.

Then what remains is more subjective:
- Should there be a body judgement? I believe it is a good thing to have but it's not necessary. The absence of body judgement can be mostly compensated by a higher value for tail judgements.
- For longer sliders with many ticks, the only factor increasing the total slider-value is the value of the ticks, and I have no strong opinion on this but I believe something around \~50 is a good value for slider ticks.
- Then, for the tail and body judgement, if we have a body judgement then remains the question of how the split between body and tail value should be and for that I believe \~100 for the body and \~100 for the tail should work fine. If there's no body then \~150 for the tail seems reasonable.

**Reply by Zyfarok** [reply]
Two additions:

- Regarding visual feedback, if you don't have a judgement displayed at the tail (as in `current`), the most tricky part is how to give feedback for slider-ends. In particular, you can leave the slider 36ms before the end and you should be able to distinguish between that and a missed slider-end.

- Regarding the value of body judgement (which I suggest to give a value of ~100 above), if we decide to keep it, we can just give 0 when you miss any tick or slider-end, and give the full value otherwise. It should not be affected by the accuracy of the head, and there's also no need for the complexity of multiple intermediate values (like we have in `v2-1` and `v2-2`) depending on a proportion of ticks hit. This judgement is only here to reward holding the key till the end and aiming all ticks and ends correctly.

**Comment by Givikap120** [comment]
Gameplay-wise v2-1 is the best, I can't see any alternatives to it
But there are some non-gameplay bad things here:
- the total amount of judgements will be different from stable which can cause some issues. However, from the mine perspective - I don't care.
- considering that judgements for tapping are mixed with judgements for sliderbody - you have lack of info about the precise amount of mistaps and the precise amount of sliderbody mistakes. As a player - I won't care, but as a pp reworks developer - this will lead to many issues. Specifically - you won't be able to punish missed sliders as much as you have to. You will either overpunish or underpunish.

**Reply by bdach** [reply]
> considering that judgements for tapping are mixed with judgements for sliderbody - you have lack of info about the precise amount of mistaps and the precise amount of sliderbody mistakes. As a player - I won't care, but as a pp reworks developer - this will lead to many issues. Specifically - you won't be able to punish missed sliders as much as you have to. You will either overpunish or underpunish.

I'm not sure this point matters as it has been stated several times that the pp algorthms have to continue to work for old scores where this data isn't there, and maintaining two code paths for this will not be allowed. So ultimately there are going to be scores that are missing this information either way.

**Reply by Givikap120** [reply]
This missed information is already guessed by two algorithms (one guess amount of missed sliderheads, second - amount of missed sliderends). So if the score has this information - it will be used instead of guessing. 
And of course, having precise information is much better. So if you want to have your score judged more precisely and correctly - you should play lazer.

**Comment by Sleepteiner** [comment]
Of the three options, I like v2-2 the best. Although, that is mostly just because it is the most punishing on the resulting accuracy. The less punishing messing up a slider is, the less players will be incentivized to play them correctly, and therefore the more negative impact there will be on the flow of thousands of ranked maps that expect the player to be incentivized to play sliders correctly.

Having a _visual_ judgement directly after clicking the slider head has always felt off to me, but I recognize how many people want that, so I can tolerate the look of it. (I am, however, against forcing slider accuracy as the new default, but that is not the topic of this discussion.)

I really want to emphasize the importance of a visual judgement at the end of a slider that some people have been dismissing in this discussion. It is completely necessary, in my opinion.

There is currently very little visual feedback for missing a slider end. This means that new players are not getting the feedback they need to understand how to play sliders well. I've played this game actively since 2014 and after playing a lot of lazer lately, even I've been developing  some bad slider habits due to how de-emphasized slider ends are in lazer, both in accuracy and in visual feedback. Having sliders in this state for this long in lazer seems to have already damaged a generation of osu player's ability to play sliders well which is pretty depressing.

There is also currently almost no visual feedback at all (correct me if I'm wrong) when getting a combo break on a slider tick if you have the combo count in the UI turned off, like many people do. Sure, the combo break sound still happens, but sometimes combo break sounds can be drowned out by the song making it hard to tell if a combo break happened or not in an intense section of the map. I've seen people complain about this specific issue in the Spotlight playlist chats before.

Worse yet is if you get a combo break on a slider tick when the combo is too low to play the combo break sound, there is almost no indication in the gameplay that you made a mistake. [This comment has links to videos that showcase how ridiculous this is](https://github.com/ppy/osu/discussions/25810#discussioncomment-7885187) (the point of the comment itself is unrelated).

**Reply by peppy** [reply]
> Having sliders in this state for this long in lazer seems to have already damaged a generation of osu player's ability to play sliders well which is pretty depressing.

what lol.

As for visual feedback, I intend to add tick hit/miss feedback (as per stable) as a separate visual. This will happen regardless of slider judgements. So you don't have to worry about that.

**Reply by Sleepteiner** [reply]
I've been reading a lot of the feedback about lazer sliders lately and I've seen multiple players who are newer and mainly play lazer say something along the lines of "having very little focus on slider ends fits 'modern' tech mapping due to how many fast sliders there are." This is a nonsense statement and I can only interpret it as lazer simply not teaching players how to play sliders well by essentially encouraging them to ignore difficult ones, therefore damaging their ability to play difficult sliders. That is what I was referring to.

But, okay, if visual feedback is in the plan for sure, then things will probably recover.
